Dictionaries with required members that need rooting break codegen

The design for dictionaries that contain rooted members is that an empty dictionary value is created and then each member is set individually. This doesn't work when the WebIDL type specifies `required`, like the `FormDataEvent`:
```
error[E0599]: no function or associated item named `empty` found for type `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit` in the current scope
   --> /Users/cybai/codespace/mozilla/servo/target/debug/build/script-59d4418593395066/out/Bindings/FormDataEventBinding.rs:857:95
    |
259 | pub struct FormDataEventInit {
    | ---------------------------- function or associated item `empty` not found for this
...
857 |             crate::dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit::empty()
    |             ----------------------------------------------------------------------------------^^^^^
    |             |
    |             function or associated item not found in `dom::bindings::codegen::Bindings::FormDataEventBinding::FormDataEventInit`
```
Addressing this probably means figuring out the precise issue that the current design is addressing.

--- a/components/script/dom/bindings/codegen/CodegenRust.py
+++ b/components/script/dom/bindings/codegen/CodegenRust.py
@@ -6425,7 +6425,8 @@ class CGDictionary(CGThing):
         mustRoot = ""
         if self.membersNeedTracing():
             mustRoot = "#[unrooted_must_root_lint::must_root]\n"
-            derive += ["Default"]
+            if not self.hasRequiredFields(self.dictionary):
+                derive += ["Default"]
 
         return (string.Template(
                 "#[derive(${derive})]\n"
@@ -6485,16 +6486,14 @@ class CGDictionary(CGThing):
         selfName = self.makeClassName(d)
         if self.membersNeedTracing():
             actualType = "RootedTraceableBox<%s>" % selfName
-            preInitial = "let mut dictionary = RootedTraceableBox::new(%s::default());\n" % selfName
-            initParent = initParent = ("dictionary.parent = %s;\n" % initParent) if initParent else ""
-            memberInits = CGList([memberInit(m, False) for m in self.memberInfo])
-            postInitial = ""
+            preInitial = "let dictionary = RootedTraceableBox::new(%s {\n" % selfName
+            postInitial = "});\n"
         else:
             actualType = selfName
             preInitial = "let dictionary = %s {\n" % selfName
             postInitial = "};\n"
-            initParent = ("parent: %s,\n" % initParent) if initParent else ""
-            memberInits = CGList([memberInit(m, True) for m in self.memberInfo])
+        initParent = ("parent: %s,\n" % initParent) if initParent else ""
+        memberInits = CGList([memberInit(m, True) for m in self.memberInfo])
 
         return string.Template(
             "impl ${selfName} {\n"
@@ -6540,8 +6539,8 @@ class CGDictionary(CGThing):
                 "initParent": CGIndenter(CGGeneric(initParent), indentLevel=16).define(),
                 "initMembers": CGIndenter(memberInits, indentLevel=16).define(),
                 "insertMembers": CGIndenter(memberInserts, indentLevel=8).define(),
-                "preInitial": CGIndenter(CGGeneric(preInitial), indentLevel=16).define(),
-                "postInitial": CGIndenter(CGGeneric(postInitial), indentLevel=16).define(),
+                "preInitial": CGIndenter(CGGeneric(preInitial), indentLevel=8).define(),
+                "postInitial": CGIndenter(CGGeneric(postInitial), indentLevel=8).define(),
             })
 
     def membersNeedTracing(self):

--- a/components/script/dom/promiserejectionevent.rs
+++ b/components/script/dom/promiserejectionevent.rs
@@ -5,7 +5,7 @@
 use crate::dom::bindings::codegen::Bindings::EventBinding::EventMethods;
 use crate::dom::bindings::codegen::Bindings::PromiseRejectionEventBinding;
 use crate::dom::bindings::codegen::Bindings::PromiseRejectionEventBinding::PromiseRejectionEventMethods;
-use crate::dom::bindings::error::{Error, Fallible};
+use crate::dom::bindings::error::Fallible;
 use crate::dom::bindings::inheritance::Castable;
 use crate::dom::bindings::reflector::reflect_dom_object;
 use crate::dom::bindings::root::DomRoot;
@@ -72,14 +72,7 @@ impl PromiseRejectionEvent {
         init: RootedTraceableBox<PromiseRejectionEventBinding::PromiseRejectionEventInit>,
     ) -> Fallible<DomRoot<Self>> {
         let reason = init.reason.handle();
-        let promise = match init.promise.as_ref() {
-            Some(promise) => promise.clone(),
-            None => {
-                return Err(Error::Type(
-                    "required member promise is undefined.".to_string(),
-                ));
-            },
-        };
+        let promise = init.promise.clone();
         let bubbles = EventBubbles::from(init.parent.bubbles);
         let cancelable = EventCancelable::from(init.parent.cancelable);
 

--- a/components/script/dom/webidls/PromiseRejectionEvent.webidl
+++ b/components/script/dom/webidls/PromiseRejectionEvent.webidl
@@ -6,12 +6,12 @@
 
 [Exposed=(Window,Worker)]
 interface PromiseRejectionEvent : Event {
-  [Throws] constructor(DOMString type, optional PromiseRejectionEventInit eventInitDict = {});
+  [Throws] constructor(DOMString type, PromiseRejectionEventInit eventInitDict);
   readonly attribute Promise<any> promise;
   readonly attribute any reason;
 };
 
 dictionary PromiseRejectionEventInit : EventInit {
-  /* required */ Promise<any> promise;
+  required Promise<any> promise;
   any reason;
 };

--- a/tests/wpt/metadata/html/dom/idlharness.https.html.ini
+++ b/tests/wpt/metadata/html/dom/idlharness.https.html.ini
@@ -1163,9 +1163,6 @@
   [OffscreenCanvasRenderingContext2D interface: operation closePath()]
     expected: FAIL
 
-  [PromiseRejectionEvent interface object length]
-    expected: FAIL
-
   [CanvasRenderingContext2D interface: document.createElement("canvas").getContext("2d") must inherit property "textAlign" with the proper type]
     expected: FAIL
 

--- a/tests/wpt/metadata/html/dom/idlharness.worker.js.ini
+++ b/tests/wpt/metadata/html/dom/idlharness.worker.js.ini
@@ -68,9 +68,6 @@
   [OffscreenCanvasRenderingContext2D interface: operation restore()]
     expected: FAIL
 
-  [PromiseRejectionEvent interface object length]
-    expected: FAIL
-
   [WorkerGlobalScope interface: calling createImageBitmap(ImageBitmapSource, ImageBitmapOptions) on self with too few arguments must throw TypeError]
     expected: FAIL
 

